Remove dependencies on non-live illink

sbomer opened the pull request.

sbomer committed:
Remove dependencies on non-live illink

sbomer committed:
Fix properties

jkoritzinsky reviewed (commented):
Can you take a look in targetingpacks.targets and update the KnownILLink item changes to point to the live ILLink?

sbomer committed:
Remove KnownILLinkPack

sbomer committed:
Merge remote-tracking branch 'origin/main' into liveILLink

sbomer committed:
Don't require illink analyzers for source generator projects

Microsoft.Extensions.Options.SourceGeneration was enabling
the single-file analyzer, but did not import illink.targets
because it is not a source project.

We could fix this by importing liveILLink.targets instead, but
this introduces more problems with circular dependencies
(generator projects depending on illink depending on
generator projects).

ViktorHofer reviewed (commented):
First of all: Amazing change!!

Slightly unrelated to this PR but still important from my point of view: Is there a feature flag that we can condition the illink.targets import on so that only project that actually use the linker, import the file and with it build the P2P to ILLink.Tasks.csproj?

ViktorHofer commented on eng/liveILLink.targets:
```diff
@@ -0,0 +1,29 @@
+  <PropertyGroup>
+    <!-- Don't use SDK's trimming functionality. -->
+    <_RequiresILLinkPack>false</_RequiresILLinkPack>
+    <ILLinkTasksAssembly>$(ToolsILLinkDir)$(NetCoreAppToolCurrent)/ILLink.Tasks.dll</ILLinkTasksAssembly>     
```
Please use a windows separator char. We standardized on using that in anything under src/libraries and that file is imported by our projects.

```suggestion
    <ILLinkTasksAssembly>$(ToolsILLinkDir)$(NetCoreAppToolCurrent)\ILLink.Tasks.dll</ILLinkTasksAssembly>     
```

ViktorHofer commented on eng/Analyzers.targets:
```diff
@@ -14,7 +14,8 @@
       '$(EnableSingleFileAnalyzer)' == '' And
-      '$(TargetFrameworkIdentifier)' == '.NETCoreApp'">true</EnableSingleFileAnalyzer>
+      '$(TargetFrameworkIdentifier)' == '.NETCoreApp' And
+      '$(IsGeneratorProject)' != 'true'">true</EnableSingleFileAnalyzer>
```
What about IsReferenceSourceProject?

ViktorHofer replied:
Yes, I don't think we should have this analyzer enabled for test projects at the moment. So, I would prefer if you keep your change.

sbomer committed:
Only enable single-file analyzer for source projects

This more closely matches the condition on the import of
illink.targets. When EnableSingleFileAnalyzer is true, we should
also import liveILLink.targets so that we get the live analyzer
bits.

Also address feedback about path separators.

sbomer committed:
Separate imports

This will make it easier to use different conditions for the two.

sbomer committed:
Fix libraries build

Some libraries test projects set PublishTrimmed, so they require
the liveILLink.targets.

sbomer committed:
Add _RequiresLiveILLink flag

sbomer committed:
Revert "Fix libraries build"

This reverts commit e83a2a06ebeadc8577c9b9d97acbcffd13e8aab2.

sbomer committed:
Revert "Separate imports"

This reverts commit 597a012bfed4a7f31e7d91a66a4c6348a8419bbf.

sbomer committed:
Set _RequiresLiveILLink in illink.targets

sbomer committed:
Don't restore KnownILLinkPack in trimming tests

sbomer committed:
Import liveILLink.targets in libraries projects

This should let libraries projects (whether they are source or test
projects) use the SDK's trimming properties with a live ILLink, enabling
the ProjectReference only for those projects that need it.

Because illink.targets also imports liveILLink.targets, this import is
guarded by a property to ensure liveILLink.targets is only imported
once on this path.

sbomer committed:
Fix import path

ViktorHofer commented on eng/Analyzers.targets:
```diff
@@ -14,7 +14,8 @@
     <EnableSingleFileAnalyzer Condition="
       '$(EnableSingleFileAnalyzer)' == '' And
-      '$(TargetFrameworkIdentifier)' == '.NETCoreApp'">true</EnableSingleFileAnalyzer>
+      '$(TargetFrameworkIdentifier)' == '.NETCoreApp' And
```
nit: I know this was already there before but please use lowercase conditional operators. Just style. I would really like msbuild to add a style analyzer so that we don't have the inconsistencies in our repo.

```suggestion
      '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and
```

ViktorHofer commented on src/libraries/Directory.Build.targets:
```diff
@@ -128,6 +128,8 @@
   <Import Project="$(RepositoryEngineeringDir)slngen.targets" Condition="'$(IsSlnGen)' == 'true'" />
 
   <Import Project="$(RepositoryEngineeringDir)illink.targets" Condition="'$(IsSourceProject)' == 'true' or '$(ExplicitlyImportCustomILLinkTargets)' == 'true'" />
+  <!-- liveILLink.targets should be conditionally imported after illink.targets (which also imports liveILLink.targets) to prevent a double-import. -->
+  <Import Project="$(RepositoryEngineeringDir)liveILLink.targets" Condition="'$(_ImportedLiveILLinkTargets)' != 'true'" />
```
Changes look really good except for these two lines (which I don't understand). illink.targets already imports liveILLink.targets. Why do we need to import it here explicitly?

sbomer committed:
Separate imports, PR feedback

Avoid the need for an import guard by requiring liveILLink.targets
to be imported separately from illink.targets.

Also fix casing of MSBuild And in Analyzers.targets

sbomer committed:
Move liveILlink.targets import to top-level Directory.Build.targets

sbomer committed:
Clean up redundant setting

sbomer committed:
Fix import ordering for nativeaot test build

sbomer committed:
Try removing imports from tests.mobile.targets

Mobile trimmingtests should already use the live illink with
the logic in project.csproj.template. The import of illink.targets
shouldn't be required, unless tests.mobile.targets are required
for other test scenarios.

This import was resulting in a double-import, and wasn't working
because liveILLink needs ToolsILLinkDir to be set, but it's not
set during the trimming test build.

sbomer committed:
Keep ILC props/items before ILC targets import

sbomer committed:
Move liveILLink.targets import back down

The top-level import was causing ordering issues in
src/libraries/Directory.Build.targets. The top-level targets
were importing liveBuilds and liveILLink, but
src/libraries/Directory.Build.targets needs to import
(indirectly through eng/testing/tests.targets)
the nativeaot targets after liveBuilds,
but before liveILLink. Moving the import of liveILLink.targets
back down allows more control of the import order.

sbomer committed:
Fix illink path

This used to use the layout of the nuget package, but now we
are using the layout of the ILLink.Tasks build output.

ViktorHofer commented:
> Assets file '/Users/runner/work/1/s/artifacts/obj/ILLink.Tasks/project.assets.json' doesn't have a target for 'net8.0/osx-x64'

sbomer committed:
Merge remote-tracking branch 'origin/main' into liveILLink

ViktorHofer committed:
Merge branch 'main' into liveILLink

sbomer committed:
Merge remote-tracking branch 'origin/main' into liveILLink

sbomer committed:
Merge remote-tracking branch 'origin/main' into liveILLink

ViktorHofer commented on src/coreclr/Directory.Build.targets:
```diff
@@ -26,7 +27,8 @@
-  <Import Condition="'$(MSBuildProjectName)' == 'System.Private.CoreLib' or '$(IsNativeAotProject)' == 'true'" Project="$(RepositoryEngineeringDir)illink.targets" />
+    <!-- Import targets here to have TargetPath and other macros defined. Limit to CoreLib. -->
+    <Import Condition="'$(MSBuildProjectName)' == 'System.Private.CoreLib' or '$(IsNativeAotProject)' == 'true'" Project="$(RepositoryEngineeringDir)illink.targets" />
+    <Import Project="$(RepositoryEngineeringDir)liveILLink.targets" />
```
Apparently this causes crossgen2_publish.proj to reference the linker projects now. These projects flow some global properties between each other which causes binclashes, race conditions and in this case, restore errors because the project.assets.json file doesn't list a required RID.

see https://github.com/dotnet/runtime/pull/92826/files#r1341932080 for more details

Can we further fine tune this import to only apply to the projects that really need the live linker?

ViktorHofer replied:
Please look at a binlog and see if all the projects that now reference the linker project actually need it. I'm not sure about the crossgen2_publish case though.

jkoritzinsky replied:
Crossgen2_publish needs the linker for platforms that don’t support NativeAOT.

ViktorHofer committed:
Merge branch 'main' into liveILLink

ViktorHofer commented on src/coreclr/Directory.Build.targets:
```diff
@@ -26,7 +27,8 @@
 
-  <!-- Import targets here to have TargetPath and other macros defined. Limit to CoreLib. -->
-  <Import Condition="'$(MSBuildProjectName)' == 'System.Private.CoreLib' or '$(IsNativeAotProject)' == 'true'" Project="$(RepositoryEngineeringDir)illink.targets" />
+    <!-- Import targets here to have TargetPath and other macros defined. Limit to CoreLib. -->
+    <Import Condition="'$(MSBuildProjectName)' == 'System.Private.CoreLib' or '$(IsNativeAotProject)' == 'true'" Project="$(RepositoryEngineeringDir)illink.targets" />
+    <Import Project="$(RepositoryEngineeringDir)liveILLink.targets" />
```
nit: indentation

```suggestion
  <!-- Import targets here to have TargetPath and other macros defined. Limit to CoreLib. -->
  <Import Condition="'$(MSBuildProjectName)' == 'System.Private.CoreLib' or '$(IsNativeAotProject)' == 'true'" Project="$(RepositoryEngineeringDir)illink.targets" />
  <Import Project="$(RepositoryEngineeringDir)liveILLink.targets" />
```

ViktorHofer commented on src/libraries/Directory.Build.targets:
```diff
@@ -127,6 +127,7 @@
   <Import Project="$(RepositoryEngineeringDir)testing\coverage.targets" Condition="'$(EnableRunSettingsSupport)' == 'true' or '$(EnableCoverageSupport)' == 'true'" />
 
   <Import Project="$(RepositoryEngineeringDir)illink.targets" Condition="'$(IsSourceProject)' == 'true' or '$(ExplicitlyImportCustomILLinkTargets)' == 'true'" />
+  <Import Project="$(RepositoryEngineeringDir)liveILLink.targets" />
```
Would it make sense to only import this file when the illink.targets file is imported?

```suggestion
  <ImportGroup Condition="'$(IsSourceProject)' == 'true' or '$(ExplicitlyImportCustomILLinkTargets)' == 'true'">
    <Import Project="$(RepositoryEngineeringDir)illink.targets" />
    <Import Project="$(RepositoryEngineeringDir)liveILLink.targets" />
  </ImportGroup>
```

ViktorHofer commented on eng/liveILLink.targets:
```diff
@@ -0,0 +1,42 @@
+  <PropertyGroup Condition="'$(_RequiresLiveILLink)' == 'true'">
+    <!-- Don't use SDK's trimming functionality. -->
+    <_RequiresILLinkPack>false</_RequiresILLinkPack>
+    <ILLinkTasksAssembly>$(ToolsILLinkDir)$(NetCoreAppToolCurrent)\ILLink.Tasks.dll</ILLinkTasksAssembly>
```
This needs to respect the MSBuildRuntimeType and pick the .NET Framework one, same as https://github.com/dotnet/runtime/blob/d3569b99ead8f909556f5b416b6f57d94f30e668/eng/illink.targets#L15-L17

ViktorHofer replied:
```suggestion
    <ILLinkTasksAssembly Condition="'$(MSBuildRuntimeType)' == 'Core'">$(ToolsILLinkDir)$(NetCoreAppToolCurrent)\ILLink.Tasks.dll</ILLinkTasksAssembly>
    <ILLinkTasksAssembly Condition="'$(MSBuildRuntimeType)' != 'Core'">$(ToolsILLinkDir)$(NetFrameworkToolCurrent)\ILLink.Tasks.dll</ILLinkTasksAssembly>
```

sbomer committed:
Apply suggestions from code review

- Fix indentation
- Only import liveILLink.targets in libraries when illink.targets is imported
- Respect MSBuildRuntimeType for live ILLinkTasksAssembly override

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

sbomer committed:
Merge remote-tracking branch 'origin/main' into liveILLink

jkoritzinsky commented on eng/targetingpacks.targets:
We should update the logic in this file that removes the PackageReference to ILCompiler to also remove the reference to the ILLink packages that can be pulled in.

sbomer committed:
Fix libs test project that sets PublishTrimmed

sbomer committed:
Revert "Fix libs test project that sets PublishTrimmed"

This reverts commit be50c3575e8b68a4769e51febc3d5c5a5421cb5b.

sbomer committed:
Revert condition on liveILLink.targets import

Various test configurations (wasm, nativeaot) rely on the live ILLink
when running libraries tests.

ViktorHofer commented:
Looks like there is a remaining failure in the NativeAOT leg .

sbomer committed:
Skip nonexistent targets when copying native binaries

This target isn't defined in the ILLink.Tasks.csproj ProjectReference.
Try skipping nonexistent targets to fix this.
